packages/iframe-execution-environment-service/src/IframeExecutionEnvironmentService.ts

```diff
@@ -0,0 +1,341 @@
+        `Plugin "${pluginData.pluginName}" is already being executed.`,
+      );
+    }
+
+    const job = await this._init();
+    this._mapPluginAndJob(pluginData.pluginName, job.id);
+    this.setupPluginProvider(
+      pluginData.pluginName,
+      job.streams.rpc as unknown as Duplex,
+    );
+
+    const result = await this._command(job.id, {
+      jsonrpc: '2.0',
+      method: 'executePlugin',
+      params: pluginData,
+      id: nanoid(),
+    });
+    this._createPluginHooks(pluginData.pluginName, job.id);
+    return result;
+  }
+
+  _mapPluginAndJob(pluginName: string, jobId: string): void {
+    this.pluginToJobMap.set(pluginName, jobId);
+    this.jobToPluginMap.set(jobId, pluginName);
+  }
+
+  /**
+   * @returns The ID of the plugin's job.
+   */
+  _getJobForPlugin(pluginName: string): string | undefined {
+    return this.pluginToJobMap.get(pluginName);
+  }
+
+  /**
+   * @returns The ID job's plugin.
+   */
+  _getPluginForJob(jobId: string): string | undefined {
+    return this.jobToPluginMap.get(jobId);
+  }
+
+  _removePluginAndJobMapping(jobId: string): void {
+    const pluginName = this.jobToPluginMap.get(jobId);
+    if (!pluginName) {
+      throw new Error(`job: "${jobId}" has no mapped plugin.`);
+    }
+
+    this.jobToPluginMap.delete(jobId);
+    this.pluginToJobMap.delete(pluginName);
+  }
+
+  private async _init(): Promise<EnvMetadata> {
+    const jobId = nanoid();
+    const streams = await this._initStreams(jobId);
+    const rpcEngine = new JsonRpcEngine();
+
+    const jsonRpcConnection = createStreamMiddleware();
+
+    pump(jsonRpcConnection.stream, streams.command, jsonRpcConnection.stream);
+
+    rpcEngine.push(jsonRpcConnection.middleware);
+
+    const envMetadata = {
+      id: jobId,
+      streams,
+      rpcEngine,
+    };
+    this._setJob(jobId, envMetadata);
+
+    await this._command(jobId, {
+      jsonrpc: '2.0',
+      method: 'handshake',
+      id: nanoid(),
+    });
+
+    return envMetadata;
+  }
+
+  async _initStreams(envId: string): Promise<any> {
+    this._iframeWindow = await this._createWindow(
+      this.iframeUrl.toString(),
+      envId,
+    );
+    const envStream = new WindowPostMessageStream({
+      name: 'parent',
+      target: 'child',
+      targetWindow: this._iframeWindow,
+    });
+    // Typecast justification: stream type mismatch
+    const mux = setupMultiplex(
+      envStream as unknown as Duplex,
+      `Environment:${envId}`,
+    );
+
+    const commandStream = mux.createStream(PLUGIN_STREAM_NAMES.COMMAND);
+    commandStream.write({
+      jsonrpc: '2.0',
+      method: 'handshake',
+      params: [],
+      id: 0,
+    });
+    const rpcStream = mux.createStream(PLUGIN_STREAM_NAMES.JSON_RPC);
+
+    // Typecast: stream type mismatch
+    return {
+      command: commandStream as unknown as Duplex,
+      rpc: rpcStream,
+      _connection: envStream,
+    };
+  }
+
+  _createWindow(uri: string, envId: string, timeout = 60000): Promise<Window> {
```

Comment:
60 seconds seems like a lot. If it ever takes nearly that long, it's game over anyway. For testing purposes, we do want to give it extra time to load though... How about 20 seconds?